src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -264,6 +303,80 @@ impl<'a, C: Client<T>, T: Timestamp + Lattice> StorageControllerMut<'a, C, T> {
             .ok_or(StorageError::IdentifierMissing(id))
     }
 
+    /// Persist timestamp bindings updates received from ingestion workers
+    pub(super) async fn persist_timestamp_bindings(
+        &mut self,
+        feedback: &TimestampBindingFeedback<T>,
+    ) {
+        for (id, bindings) in &feedback.bindings {
+            let mut ts_binding_stash = self
+                .storage
+                .stash
+                .collection::<PartitionId, ()>(&id.to_string())
+                .unwrap();
+
+            let collection_state = self.collection_mut(*id).expect("misisng_source_id");
+
+            let upper = ts_binding_stash.upper().unwrap();
+            let since = ts_binding_stash.since().unwrap();
+            let mut seal_frontier = upper.clone();
+
+            // Here we differentialize the bindings we got from workers
+            ts_binding_stash
+                .update_many(bindings.iter().cloned().flat_map(|(pid, ts, offset)| {
+                    let prev_offset = collection_state
+                        .last_reported_ts_bindings
+                        .entry(pid.clone())
+                        .or_default();
+
+                    let ts = ts.try_into().unwrap();
+                    let update = ((pid, ()), ts, offset.offset - prev_offset.offset);
+
+                    if seal_frontier.less_than(&ts) {
+                        seal_frontier = Max1Antichain::Singular(ts);
```

Comment:
The whole `Max1Antichain` thing continues to confuse me. It looks like `Stash` made that call, but .. yeah.

Reply:
What would you expect to see here with a normal antichain? Can we add that same API to `Max1Antichain`?

Reply:
That's what would be clearest. It's possible that I don't have the `Stash` requirements top of mind, but to the extent that `Max1Antichain` is a perf optimization, meh. It seemed like if it was an "I can be converted to `{ number, NULL }`" then it seemed that it could be a wrapper around `Antichain` with a `T: TotalOrder` constraint.

Not sure, but my opinions on "roll your own" TD/DD stuff has evolved. :D No rush to change things that work, but I end up needing to re-understand if it does what I think when it is a new type to me.

Reply:
I suppose it's sort of a perf optimization, but yeah, I'd phrase the requirement as restricting the antichain to a type we know we can represent in SQL. 

> then it could be a wrapper around Antichain with a T: TotalOrder constraint.

That sounds right! I think the Frank that reviewed the stash PR hadn't come up with that yet. :D